File: 0050-Pow-xn/lc0050.cc
// C++ solution.

class Solution {
public:
  double myPow(double x, int n) {
    if (n == std::numeric_limits<int>::lowest()){
      return myPow(1./x, -(n+1)) / x;
    }
    if (n < 0) {
      return myPow(1./x, -n);
    }

    double res = 1.;
    for (; n; n >>= 1) {
      if (n&1) { res *= x; }
      x *= x;
    }
    return res;
  }
};

File: 0050-Pow-xn/lc0050.rs
// Rust solution.

impl Solution {
    pub fn my_pow(x: f64, n: i32) -> f64 {
        match n {
            std::i32::MIN => Self::my_pow(1_f64/x, -(n+1)) / x,
            _ if n < 0 => Self::my_pow(1_f64/x, -n),
            _ => {
                let mut res: f64 = 1.0;
                let (mut x, mut n) = (x, n);
                while n > 0 {
                    if n&1 == 1 { res *= x; }
                    x *= x;
                    n = n >> 1;
                }
                res
            }
        }
    }
}
